panoply/datasource.py

```diff
@@ -18,13 +21,58 @@ def log(self, *msgs):
         if "logger" in self.options:
             self.options["logger"](msgs)
         else:
-            print msgs
+            print(msgs)
 
-    def progress(self, loaded, total, msg):
+    def progress(self, loaded, total, msg=''):
         self.fire("progress", {
             "loaded": loaded,
             "total": total,
             "msg": msg
         })
 
+# a decorator used to invalidate the access_token for oauth based data sources
+# this should be used on every method in the data source that fetches data
+# from the server (controlled by this oauth), and that needs to be invalidated
+def invalidate_token(refresh_url, callback=None,
+                     keys=['access_token', 'refresh_token']):
+    def _invalidate_token(f):
+        def wrapper(*args):
+            self = args[0]
+            try:
+                return f(*args)
+            except Exception, e:
+                try:
+                    self.log("Reinvalidating access_token...")
+                    self.source[keys[0]] = None
+
+                    # get a new token from refresh_url
+                    token = self.source.get(keys[1])
+                    r = requests.post(
+                        refresh_url,
+                        data = dict(self.options['refresh'], **{keys[1]:token}))
```

Comment:
This is the same question as you asked before - what if the keys list is empty.
We set some guidelines as to how one should write a DS in order to get the functionality he/she wants. If the developer doesn't follow those guidelines, he will get an error. We shouldn't validate everything ...